iot/end_to_end_example/device/app/src/main/kotlin/com/example/androidthings/fancontrol/FanControlActivity.kt

```diff
@@ -0,0 +1,379 @@
+            for (i in mRainbow.indices) {
+                mRainbow[i] = Color.rgb(0, 0, 0)
+            }
+        } catch (e: IOException) {
+            mLedstrip = null // Led strip is optional.
+        }
+
+        mHandler.post(mAnimateRunnable)
+
+        // GPIO led
+        try {
+            val pioManager = PeripheralManager.getInstance()
+            mLed = pioManager.openGpio(BoardDefaults.ledGpioPin)
+            mLed!!.setEdgeTriggerType(Gpio.EDGE_NONE)
+            mLed!!.setDirection(Gpio.DIRECTION_OUT_INITIALLY_LOW)
+            mLed!!.setActiveType(Gpio.ACTIVE_HIGH)
+        } catch (e: IOException) {
+            throw RuntimeException("Error initializing led", e)
+        }
+
+        // Configure the Cloud IoT Connector --
+        val pkId = resources.getIdentifier("privatekey", "raw", packageName)
+        try {
+            if (pkId != 0) {
+                val privateKey = applicationContext
+                        .resources.openRawResource(pkId)
+                val keyBytes = inputStreamToBytes(privateKey)
+
+                val spec = PKCS8EncodedKeySpec(keyBytes)
+                val kf = KeyFactory.getInstance("RSA")
+                val keys = KeyPair(null, kf.generatePrivate(spec))
+
+                // Configure Cloud IoT Core project information
+                val connectionParams = ConnectionParams.Builder()
+                        .setProjectId(BuildConfig.PROJECT_ID)
+                        .setRegistry(BuildConfig.REGISTRY_ID, BuildConfig.ZONE)
+                        .setDeviceId(BuildConfig.DEVICE_ID)
+                        .build()
+
+                // Initialize the IoT Core client
+                client = IotCoreClient.Builder()
+                        .setConnectionParams(connectionParams)
+                        .setKeyPair(keys)
+                        .setOnConfigurationListener(OnConfigurationListener { this.onConfigurationReceived(it) })
+                        .build()
+
+                // Connect to Cloud IoT Core
+                client!!.connect()
+
+                mHandler.post(mTempReportRunnable)
+            }
+        } catch (ikse: InvalidKeySpecException) {
+            Log.e(TAG, "INVALID Key spec", ikse)
+        } catch (nsae: NoSuchAlgorithmException) {
+            Log.e(TAG, "Algorithm not supported", nsae)
+        } catch (ioe: IOException) {
+            Log.e(TAG, "Could not load key from file", ioe)
+        }
+    }
+
+    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
+        if (keyCode == KeyEvent.KEYCODE_A) {
+            Log.d(TAG, "A Click")
+            return true
+        }
+        return super.onKeyUp(keyCode, event)
+    }
+
+    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
+        return if (keyCode == KeyEvent.KEYCODE_A) {
+            true
+        } else super.onKeyUp(keyCode, event)
+    }
+
+    override fun onDestroy() {
+        super.onDestroy()
```

Comment:
shouldn't onDestroy to go last? destroy current class and then the subclass.

it's reverse of onCreate, etc.